### PR TITLE
 Make datastore generics use lookup type 

### PR DESCRIPTION
**Pull request by vidartf**
Alternative to consider.

### CODE REVIEW

**Comment by vidartf** [comment]
Note: I don't think that the "precompute" of the schema map as a generic arg is completely necessary (but it might be, not sure). I didn't try it out too much because I thought it was nicer to not have `[number]` all over 😄 Open for arguments the other way if they are good ;)

**Comment by vidartf** [comment]
CI failure is because docs build are still using 3.x of phosphor and these changes require 3.4 at least.

**Comment by saulshanabrook** [comment]
@vidartf could you give an example where this gives better type analysis than the original PR?

File: packages/datastore/tests/src/datastore.spec.ts [hunk 1]
```diff
@@ -198,10 +198,15 @@ describe('@phosphor/datastore', () => {
           expect(change.storeId).to.equal(DATASTORE_ID);
           expect(change.change['test-schema-1']).to.not.be.undefined;
           expect(change.change['test-schema-2']).to.be.undefined;
+          // This code id mostly for checking correcting typing
+          // for change events:
+          const data = change.change['test-schema-1']!['my-record']!;
+          expect(data.visible!.previous).to.equal(false);
+          expect(data.visible!.current).to.equal(true);
```
Comment (vidartf) [inline packages/datastore/tests/src/datastore.spec.ts:205]: @saulshanabrook This bit! Without the typings for the change/patch structures I get errors like this:

```ts
Property 'visible' does not exist on type:
Change<
  TestSchema | {
    id: "test-schema-2";
    fields: {
      enabled: RegisterField<boolean>;
      count: RegisterField<number>;
      keywords: ListField<string>;
    };
  }
>
```

In other words, the `change.change['test-schema-1']` line isn't able to narrow it to `Change<TestSchema>`. Note that the test here was improved a bit to highlight this issue (the two schemas now have different types).

Comment (saulshanabrook) [reply packages/datastore/tests/src/datastore.spec.ts:205]: Ah great! Thank you. 

That makes sense, so we have to tie the ID on the change to the schema.

Comment (vidartf) [reply packages/datastore/tests/src/datastore.spec.ts:205]: Yeah, and the cleanest way I was able to find were with these guys: https://github.com/saulshanabrook/phosphor/pull/1/files#diff-e8ef547444f515352fab0946e93e81e5R648-R676

**Pull request closed without merging**